x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -790,21 +789,27 @@ private Expression simplify(BinaryComparison bc) {
         @Override
         protected LogicalPlan rule(LogicalPlan p) {
             return p.transformExpressionsDown(e -> {
-                if (e instanceof Min || e instanceof Max || e instanceof Avg || e instanceof Sum) {
-                    NumericAggregate a = (NumericAggregate) e;
+                if (e instanceof Min || e instanceof Max || e instanceof Avg || e instanceof Sum ||
+                    (e instanceof Count && ((Count) e).distinct())) {
+
+                    AggregateFunction a = (AggregateFunction) e;
 
                     if (a.field().foldable()) {
                         Expression countOne = new Count(a.source(), new Literal(Source.EMPTY, 1, a.dataType()), false);
                         Equals countEqZero = new Equals(a.source(), countOne, new Literal(Source.EMPTY, 0, a.dataType()));
                         Expression argument = a.field();
                         Literal foldedArgument = new Literal(argument.source(), argument.fold(), a.dataType());
 
+                        Expression iffResult = Literal.NULL;
                         Expression iifElseResult = foldedArgument;
                         if (e instanceof Sum) {
                             iifElseResult = new Mul(a.source(), countOne, foldedArgument);
+                        } else if (e instanceof Count) {
```

Comment:
Considering there's a dedicated `Count` rule for local relationship, this snippet of code should sit there (`ReplaceCountInLocalRelation`) not here.

Reply:
This doesn't fix only for local relations:
```
SELECT COUNT(1), COUNT(DISTINCT 1) FROM test_emp;
```
should return 100, 1
```
SELECT COUNT(1), COUNT(DISTINCT 1), gender FROM test_emp;
```
should return:
```
1
1
1
```
for the distinct column.

Reply:
```
psql (12.2)

matriv=# select count(1), count(distinct 1) from test_emp;
 count | count 
-------+-------
   100 |     1
(1 row)

matriv=# select count(1), count(distinct 1), gender from test_emp group by gender;
 count | count | gender 
-------+-------+--------
    33 |     1 | F
    57 |     1 | M
    10 |     1 | 
(3 rows)
```